backend/src/services/auto_trader.py

```
@@ -33,238 +23,431 @@ def __init__(self, socketio=None, data_fetcher_service=None, ai_analysis_service
-                    logger.warning(f"AutoTrader: Stop-loss triggered for {details['symbol']}. Selling...")
-                    # Execute sell order (sell all owned tokens of this type)
-                    sell_amount = details["current_amount_tokens"] if details["current_amount_tokens"] > 0 else 0.000001 # Placeholder
+                        if sell_percentage >= 1.0:
+                            sell_amount = current_balance
+                        else:
+                            sell_amount = initial_amount * sell_percentage
+                            # Ensure we don't try to sell more than we have
+                            sell_amount = min(sell_amount, current_balance)
+
+                        hit_tiers.append(target_x)
+                        if 'metadata' not in details: details['metadata'] = {}
```

Ruff (0.15.12)

[error] Multiple statements on one line (colon)

(E701)